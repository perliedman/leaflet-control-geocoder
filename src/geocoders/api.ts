import * as L from 'leaflet';

/**
 * An object that represents a result from a geocoding query
 */
export interface GeocodingResult {
  /**
   * Name of found location
   */
  name: string;
  /**
   * The bounds of the location
   */
  bbox: L.LatLngBounds;
  /**
   * The center coordinate of the location
   */
  center: L.LatLng;
  /**
   * URL for icon representing result; optional
   */
  icon?: string;
  /**
   * HTML formatted representation of the name
   */
  html?: string;
  /**
   * Additional properties returned by the geocoder
   */
  properties?: any;
}

/**
 * A callback function used in {@link IGeocoder.geocode} and {@link IGeocoder.suggest} and {@link IGeocoder.reverse}
 */
export type GeocodingCallback = (result: GeocodingResult[]) => void;

/**
 * An interface implemented to respond to geocoding queries
 */
export interface IGeocoder {
  /**
   * Performs a geocoding query and returns the results to the callback in the provided context
   * @param query the query
   * @param cb the callback function
   * @param context the `this` context in the callback
   */
  geocode(query: string, cb: GeocodingCallback, context?: any): void;
  /**
   * Performs a geocoding query suggestion (this happens while typing) and returns the results to the callback in the provided context
   * @param query the query
   * @param cb the callback function
   * @param context the `this` context in the callback
   */
  suggest?(query: string, cb: GeocodingCallback, context?: any): void;
  /**
   * Performs a reverse geocoding query and returns the results to the callback in the provided context
   * @param location the coordinate to reverse geocode
   * @param scale the map scale possibly used for reverse geocoding
   * @param cb the callback function
   * @param context the `this` context in the callback
   */
  reverse?(location: L.LatLngLiteral, scale: number, cb: GeocodingCallback, context?: any): void;
}

export interface GeocoderOptions {
  /**
   * URL of the service
   */
  serviceUrl: string;
  /**
   * Additional URL parameters (strings) that will be added to geocoding requests
   */
  geocodingQueryParams?: Record<string, unknown>;
  /**
   * Additional URL parameters (strings) that will be added to reverse geocoding requests
   */
  reverseQueryParams?: Record<string, unknown>;
  /**
   * API key to use this service
   */
  apiKey?: string;
}

/**
 * @internal
 */
export function geocodingParams(
  options: GeocoderOptions,
  params: Record<string, unknown>
): Record<string, unknown> {
  return L.Util.extend(params, options.geocodingQueryParams);
}

/**
 * @internal
 */
export function reverseParams(
  options: GeocoderOptions,
  params: Record<string, unknown>
): Record<string, unknown> {
  return L.Util.extend(params, options.reverseQueryParams);
}
