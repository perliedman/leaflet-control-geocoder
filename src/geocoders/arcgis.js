import L from 'leaflet';
import { getJSON } from '../util';

export default {
  class: L.Class.extend({
    options: {
      service_url: 'http://geocode.arcgis.com/arcgis/rest/services/World/GeocodeServer'
    },

    initialize: function(accessToken, options) {
      L.setOptions(this, options);
      this._accessToken = accessToken;
    },

    geocode: function(query, cb, context) {
      var params = {
        SingleLine: query,
        outFields: 'Addr_Type',
        forStorage: false,
        maxLocations: 10,
        f: 'json'
      };

      if (this._key && this._key.length) {
        params.token = this._key;
      }

      getJSON(this.options.service_url + '/findAddressCandidates', params, function(data) {
        var results = [],
          loc,
          latLng,
          latLngBounds;

        if (data.candidates && data.candidates.length) {
          for (var i = 0; i <= data.candidates.length - 1; i++) {
            loc = data.candidates[i];
            latLng = L.latLng(loc.location.y, loc.location.x);
            latLngBounds = L.latLngBounds(
              L.latLng(loc.extent.ymax, loc.extent.xmax),
              L.latLng(loc.extent.ymin, loc.extent.xmin)
            );
            results[i] = {
              name: loc.address,
              bbox: latLngBounds,
              center: latLng
            };
          }
        }

        cb.call(context, results);
      });
    },

    suggest: function(query, cb, context) {
      return this.geocode(query, cb, context);
    },

    reverse: function(location, scale, cb, context) {
      var params = {
        location: encodeURIComponent(location.lng) + ',' + encodeURIComponent(location.lat),
        distance: 100,
        f: 'json'
      };

      getJSON(this.options.service_url + '/reverseGeocode', params, function(data) {
        var result = [],
          loc;

        if (data && !data.error) {
          loc = L.latLng(data.location.y, data.location.x);
          result.push({
            name: data.address.Match_addr,
            center: loc,
            bounds: L.latLngBounds(loc, loc)
          });
        }

        cb.call(context, result);
      });
    }
  }),

  factory: function(accessToken, options) {
    return new L.Control.Geocoder.ArcGis(accessToken, options);
  }
};
