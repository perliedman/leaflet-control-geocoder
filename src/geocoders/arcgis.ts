import * as L from 'leaflet';
import { getJSON } from '../util';
import {
  IGeocoder,
  GeocoderOptions,
  GeocodingCallback,
  geocodingParams,
  GeocodingResult,
  reverseParams
} from './api';

export interface ArcGisOptions extends GeocoderOptions {}

/**
 * Implementation of the [ArcGIS geocoder](https://developers.arcgis.com/features/geocoding/)
 */
export class ArcGis implements IGeocoder {
  options: ArcGisOptions = {
    serviceUrl: 'https://geocode.arcgis.com/arcgis/rest/services/World/GeocodeServer',
    apiKey: ''
  };

  constructor(options?: Partial<ArcGisOptions>) {
    L.Util.setOptions(this, options);
  }

  geocode(query: string, cb: GeocodingCallback, context?: any): void {
    const params = geocodingParams(this.options, {
      token: this.options.apiKey,
      SingleLine: query,
      outFields: 'Addr_Type',
      forStorage: false,
      maxLocations: 10,
      f: 'json'
    });

    getJSON(this.options.serviceUrl + '/findAddressCandidates', params, data => {
      const results: GeocodingResult[] = [];
      if (data.candidates && data.candidates.length) {
        for (let i = 0; i <= data.candidates.length - 1; i++) {
          const loc = data.candidates[i];
          const latLng = L.latLng(loc.location.y, loc.location.x);
          const latLngBounds = L.latLngBounds(
            L.latLng(loc.extent.ymax, loc.extent.xmax),
            L.latLng(loc.extent.ymin, loc.extent.xmin)
          );
          results[i] = {
            name: loc.address,
            bbox: latLngBounds,
            center: latLng
          };
        }
      }

      cb.call(context, results);
    });
  }

  suggest(query: string, cb: GeocodingCallback, context?: any): void {
    return this.geocode(query, cb, context);
  }

  reverse(location: L.LatLngLiteral, scale: number, cb: GeocodingCallback, context?: any): void {
    const params = reverseParams(this.options, {
      location: encodeURIComponent(location.lng) + ',' + encodeURIComponent(location.lat),
      distance: 100,
      f: 'json'
    });
    getJSON(this.options.serviceUrl + '/reverseGeocode', params, data => {
      const result: GeocodingResult[] = [];
      if (data && !data.error) {
        const center = L.latLng(data.location.y, data.location.x);
        const bbox = L.latLngBounds(center, center);
        result.push({
          name: data.address.Match_addr,
          center: center,
          bbox: bbox
        });
      }

      cb.call(context, result);
    });
  }
}

/**
 * [Class factory method](https://leafletjs.com/reference.html#class-class-factories) for {@link ArcGis}
 * @param options the options
 */
export function arcgis(options?: Partial<ArcGisOptions>) {
  return new ArcGis(options);
}
