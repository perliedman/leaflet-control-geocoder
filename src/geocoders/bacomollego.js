var L = require('leaflet'),
	Util = require('../util'),
	proj4 = require('proj4');

module.exports = {
	class: L.Class.extend({
		options: {
			tmerc: new proj4.Proj("+proj=tmerc +lat_0=-34.629269 +lon_0=-58.463300 +k=0.999998 +x_0=100000 +y_0=100085 +ellps=intl +units=m +no_defs"),
			wsg: new proj4.Proj("EPSG:4326"),
			geocodeUrl: 'https://ws.usig.buenosaires.gob.ar/geocoder/2.2/geocoding/',
			reverseUrl: 'https://ws.usig.buenosaires.gob.ar/geocoder/2.2/reversegeocoding/',
			normalizarUrl: 'https://servicios.usig.buenosaires.gob.ar/normalizar',
			timeout: 10000,
			maxOptions: 3
		},

		initialize: function(options) {
			L.setOptions(this, options);
		},

		_normalizar: function(query, cb, context){
			Util.jsonp(this.options.normalizarUrl, {
				maxOptions: this.options.maxOptions,
				direccion: query
			}, function(json){
				cb.call(context, json);
			}, this, "callback");
			return this;
		},

		_tmerc2LatLng: function(a){
			var b = proj4.toPoint(a);
			return proj4.transform(this.options.tmerc, this.options.wsg, b),
			[ b.y, b.x ];
		},

		geocode: function(query, cb, context) {
			var results = [],
				maxOptions = 0,
				timedOut = false,
				timer;

			function checkIfCb(){
				if(!timedOut && results.length >= maxOptions){
					timedOut = true;
					cb.call(context || cb, results);
					clearTimeout(timer);
				}
			}

			function pushAndCb(c){
				results.push(c);
				checkIfCb();
			}

			this._normalizar(query, function(json){
				timer = setTimeout(L.bind(function(){
					timedOut = true;
					cb.call(context || cb, results);
				}, this), this.options.timeout);

				maxOptions = json.direccionesNormalizadas.length;

				json.direccionesNormalizadas.forEach(function(e){
					if(e.coordenadas){
						var center = L.latLng(e.coordenadas.y, e.coordenadas.x);
						pushAndCb({
							bbox: L.latLngBounds(center, center),
							name: e.direccion,
							center: center
						});
					}else if([
							(e.nombre_localidad||"").toUpperCase(),
							(e.nombre_partido||"").toUpperCase(),
					].indexOf("CABA") > -1)
						L.bind(function(e){
							Util.jsonp(this.options.geocodeUrl, {
								cod_calle: e.cod_calle,
								altura: e.altura
							}, function(tmerc){
								if(typeof(tmerc.x) != "undefined" && typeof(tmerc.y) != "undefined"){
									var center = L.latLng.apply(L, this._tmerc2LatLng([tmerc.x, tmerc.y]));
									pushAndCb({
										bbox: L.latLngBounds(center, center),
										name: e.direccion,
										center: center
									});
								}else{
									maxOptions--;
									checkIfCb();
								}
							}, this, "callback");
						}, this)(e);
					else
						this._normalizar(e.direccion, function(json){
							var data = json.direccionesNormalizadas[0];
							if(data && data.coordenadas){
								var center = L.latLng(data.coordenadas.y, data.coordenadas.x);
								pushAndCb({
									bbox: L.latLngBounds(center, center),
									name: data.direccion,
									center: center
								});
							} else{
								maxOptions--;
								checkIfCb();
							}
						}, this);
				}, this);
			}, this);
		},

		suggest: function(query, cb, context) {
			return this.geocode(query, cb, context);
		},

		reverse: function(location, scale, cb, context) {
			Util.jsonp(this.options.reverseUrl, {
				x: location.lng,
				y: location.lat,
			}, function(json){
				var results = [{
					bbox: L.latLngBounds(location, location),
					name: json.puerta,
					center: location
				}];
				cb.call(context || cb, results);
			}, "callback");
		}
	}),

	factory: function(options) {
		return new L.Control.Geocoder.BAComoLlego(options);
	}
};

