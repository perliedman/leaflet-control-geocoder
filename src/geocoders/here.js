import L from 'leaflet';
import { getJSON } from '../util';
export var HERE = L.Class.extend({
  options: {
    geocodeUrl: 'https://geocoder.api.here.com/6.2/geocode.json',
    reverseGeocodeUrl: 'https://reverse.geocoder.api.here.com/6.2/reversegeocode.json',
    app_id: '<insert your app_id here>',
    app_code: '<insert your app_code here>',
    geocodingQueryParams: {},
    reverseQueryParams: {},
    reverseGeocodeProxRadius: null
  },
  initialize: function(options) {
    L.setOptions(this, options);
  },
  geocode: function(query, cb, context) {
    var params = {
      searchtext: query,
      gen: 9,
      app_id: this.options.app_id,
      app_code: this.options.app_code,
      jsonattributes: 1
    };
    params = L.Util.extend(params, this.options.geocodingQueryParams);
    this.getJSON(this.options.geocodeUrl, params, cb, context);
  },
  reverse: function(location, scale, cb, context) {
    var _proxRadius = this.options.reverseGeocodeProxRadius
      ? this.options.reverseGeocodeProxRadius
      : null;
    var proxRadius = _proxRadius ? ',' + encodeURIComponent(_proxRadius) : '';
    var params = {
      prox: encodeURIComponent(location.lat) + ',' + encodeURIComponent(location.lng) + proxRadius,
      mode: 'retrieveAddresses',
      app_id: this.options.app_id,
      app_code: this.options.app_code,
      gen: 9,
      jsonattributes: 1
    };
    params = L.Util.extend(params, this.options.reverseQueryParams);
    this.getJSON(this.options.reverseGeocodeUrl, params, cb, context);
  },
  getJSON: function(url, params, cb, context) {
    getJSON(url, params, function(data) {
      var results = [],
        loc,
        latLng,
        latLngBounds;
      if (data.response.view && data.response.view.length) {
        for (var i = 0; i <= data.response.view[0].result.length - 1; i++) {
          loc = data.response.view[0].result[i].location;
          latLng = L.latLng(loc.displayPosition.latitude, loc.displayPosition.longitude);
          latLngBounds = L.latLngBounds(
            L.latLng(loc.mapView.topLeft.latitude, loc.mapView.topLeft.longitude),
            L.latLng(loc.mapView.bottomRight.latitude, loc.mapView.bottomRight.longitude)
          );
          results[i] = {
            name: loc.address.label,
            bbox: latLngBounds,
            center: latLng
          };
        }
      }
      cb.call(context, results);
    });
  }
});
export function here(options) {
  return new HERE(options);
}
