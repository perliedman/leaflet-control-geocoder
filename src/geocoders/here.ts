import * as L from 'leaflet';
import { getJSON } from '../util';
import { GeocoderAPI, GeocodingCallback } from './interfaces';

export interface HereOptions {
  geocodeUrl: string;
  reverseGeocodeUrl: string;
  app_id: string;
  app_code: string;
  geocodingQueryParams?: Record<string, unknown>;
  reverseQueryParams?: Record<string, unknown>;
  reverseGeocodeProxRadius: null;
}

export class HERE implements GeocoderAPI {
  options: HereOptions = {
    geocodeUrl: 'https://geocoder.api.here.com/6.2/geocode.json',
    reverseGeocodeUrl: 'https://reverse.geocoder.api.here.com/6.2/reversegeocode.json',
    app_id: '<insert your app_id here>',
    app_code: '<insert your app_code here>',
    geocodingQueryParams: {},
    reverseQueryParams: {},
    reverseGeocodeProxRadius: null
  };

  constructor(options?: Partial<HereOptions>) {
    L.Util.setOptions(this, options);
  }

  geocode(query: string, cb: GeocodingCallback, context?: any): void {
    var params = {
      searchtext: query,
      gen: 9,
      app_id: this.options.app_id,
      app_code: this.options.app_code,
      jsonattributes: 1
    };
    params = L.Util.extend(params, this.options.geocodingQueryParams);
    this.getJSON(this.options.geocodeUrl, params, cb, context);
  }

  reverse(
    location: L.LatLngLiteral,
    scale: number,
    cb: (result: any) => void,
    context?: any
  ): void {
    var _proxRadius = this.options.reverseGeocodeProxRadius
      ? this.options.reverseGeocodeProxRadius
      : null;
    var proxRadius = _proxRadius ? ',' + encodeURIComponent(_proxRadius) : '';
    var params = {
      prox: encodeURIComponent(location.lat) + ',' + encodeURIComponent(location.lng) + proxRadius,
      mode: 'retrieveAddresses',
      app_id: this.options.app_id,
      app_code: this.options.app_code,
      gen: 9,
      jsonattributes: 1
    };
    params = L.Util.extend(params, this.options.reverseQueryParams);
    this.getJSON(this.options.reverseGeocodeUrl, params, cb, context);
  }

  getJSON(url: string, params: any, cb: (result: any) => void, context?: any) {
    getJSON(url, params, function(data) {
      var results = [],
        loc,
        latLng,
        latLngBounds;
      if (data.response.view && data.response.view.length) {
        for (var i = 0; i <= data.response.view[0].result.length - 1; i++) {
          loc = data.response.view[0].result[i].location;
          latLng = L.latLng(loc.displayPosition.latitude, loc.displayPosition.longitude);
          latLngBounds = L.latLngBounds(
            L.latLng(loc.mapView.topLeft.latitude, loc.mapView.topLeft.longitude),
            L.latLng(loc.mapView.bottomRight.latitude, loc.mapView.bottomRight.longitude)
          );
          results[i] = {
            name: loc.address.label,
            properties: loc.address,
            bbox: latLngBounds,
            center: latLng
          };
        }
      }
      cb.call(context, results);
    });
  }
}

export function here(options?: Partial<HereOptions>) {
  return new HERE(options);
}
