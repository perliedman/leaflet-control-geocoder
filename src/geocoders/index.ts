export * from './api';
export * from './arcgis';
export * from './bing';
export * from './google';
export * from './here';
export * from './latlng';
export * from './mapbox';
export * from './mapquest';
export * from './neutrino';
export * from './nominatim';
export * from './open-location-code';
export * from './opencage';
export * from './pelias';
export * from './photon';
export * from './what3words';
