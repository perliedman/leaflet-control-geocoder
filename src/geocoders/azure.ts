import * as L from 'leaflet';
import { getJSON } from '../util';
import { GeocodingResult, IGeocoder } from './api';

export interface AzureMapsOptions {
  apiKey: string; // Azure Maps API Key
  serviceUrl: string; // Optional: Base URL for the Azure Maps API
}

/**
 * Implementation of [Azure Maps Geocoding](https://www.microsoft.com/en-us/maps/azure/location-services/geocoding)
 *
 * https://learn.microsoft.com/en-us/rest/api/maps/search?view=rest-maps-1.0
 */
export class AzureMaps implements IGeocoder {
  private options: AzureMapsOptions = {
    apiKey: '',
    serviceUrl: 'https://atlas.microsoft.com/search'
  };

  constructor(options: Partial<AzureMapsOptions>) {
    L.Util.setOptions(this, options);
    if (!this.options.apiKey) {
      throw new Error('Azure Maps Geocoder requires an API key.');
    }
  }

  /**
   * {@inheritdoc}
   * https://learn.microsoft.com/en-us/rest/api/maps/search/get-search-address?view=rest-maps-1.0&tabs=HTTP
   */
  async geocode(query: string): Promise<GeocodingResult[]> {
    const params = {
      'api-version': '1.0',
      query,
      'subscription-key': this.options.apiKey
    };
    const url = this.options.serviceUrl + '/address/json';
    const data = await getJSON<AzureMapsResponse>(url, params);

    return (data.results || []).map(
      (result): GeocodingResult => ({
        name: result.address.freeformAddress,
        bbox: new L.LatLngBounds(
          [result.viewport.topLeftPoint.lat, result.viewport.topLeftPoint.lon],
          [result.viewport.btmRightPoint.lat, result.viewport.btmRightPoint.lon]
        ),
        center: new L.LatLng(result.position.lat, result.position.lon)
      })
    );
  }

  /**
   * {@inheritdoc}
   * https://learn.microsoft.com/en-us/rest/api/maps/search/get-search-address-reverse?view=rest-maps-1.0&tabs=HTTP
   */
  async reverse(location: L.LatLngLiteral, scale: number): Promise<GeocodingResult[]> {
    const params = {
      'api-version': '1.0',
      query: location.lat + ',' + location.lng,
      'subscription-key': this.options.apiKey
    };
    const url = this.options.serviceUrl + '/address/reverse/json';
    const data = await getJSON<any>(url, params);

    return (data.addresses || []).map(
      (address): GeocodingResult => ({
        name: address.address.freeformAddress,
        bbox: new L.LatLngBounds(
          [address.viewport.topLeftPoint.lat, address.viewport.topLeftPoint.lon],
          [address.viewport.btmRightPoint.lat, address.viewport.btmRightPoint.lon]
        ),
        center: new L.LatLng(location.lat, location.lng)
      })
    );
  }
}

/**
 * [Class factory method](https://leafletjs.com/reference.html#class-class-factories) for {@link Azure}
 * @param options the options
 */
export function azure(options: AzureMapsOptions) {
  return new AzureMaps(options);
}

/**
 * @internal
 */
export interface AzureMapsResponse {
  summary: Summary;
  results: Result[];
}

interface Result {
  type: string;
  id: string;
  score: number;
  address: Address;
  position: Position;
  viewport: Viewport;
  entryPoints: EntryPoint[];
}

interface Address {
  streetNumber: string;
  streetName: string;
  municipalitySubdivision: string;
  municipality: string;
  countrySecondarySubdivision: string;
  countryTertiarySubdivision: string;
  countrySubdivisionCode: string;
  postalCode: string;
  extendedPostalCode: string;
  countryCode: string;
  country: string;
  countryCodeISO3: string;
  freeformAddress: string;
  countrySubdivisionName: string;
}

interface EntryPoint {
  type: string;
  position: Position;
}

interface Position {
  lat: number;
  lon: number;
}

interface Viewport {
  topLeftPoint: Position;
  btmRightPoint: Position;
}

interface Summary {
  query: string;
  queryType: string;
  queryTime: number;
  numResults: number;
  offset: number;
  totalResults: number;
  fuzzyLevel: number;
}
