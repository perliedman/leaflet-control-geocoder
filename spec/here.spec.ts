import { afterEach, describe, expect, it, vi } from 'vitest';
import { mockFetchRequest } from './mockFetchRequest';
import { HERE } from '../src/geocoders/here';
import { HEREv2 } from '../src/geocoders/here';
import { GeocodingResult } from '../src/geocoders/api';

describe('L.Control.Geocoder.HERE', () => {
  afterEach(() => vi.clearAllMocks());
  it('geocodes Innsbruck', async () => {
    const geocoder = new HERE({ app_id: 'xxx', app_code: 'yyy' });
    const result = await mockFetchRequest(
      'https://geocoder.api.here.com/6.2/geocode.json?searchtext=Innsbruck&gen=9&app_id=xxx&app_code=yyy&jsonattributes=1&maxresults=5',
      {
        response: {
          view: [
            {
              result: [
                {
                  relevance: 1,
                  matchLevel: 'city',
                  matchQuality: {
                    city: 1
                  },
                  location: {
                    locationId: 'NT_Q9dJCLiAU-LWKKq1nkKnGD',
                    locationType: 'area',
                    displayPosition: {
                      latitude: 47.268,
                      longitude: 11.3913
                    },
                    navigationPosition: [
                      {
                        latitude: 47.268,
                        longitude: 11.3913
                      }
                    ],
                    mapView: {
                      topLeft: {
                        latitude: 47.35922,
                        longitude: 11.30194
                      },
                      bottomRight: {
                        latitude: 47.21082,
                        longitude: 11.45587
                      }
                    },
                    address: {
                      label: 'Innsbruck, Tirol, Österreich',
                      country: 'AUT',
                      state: 'Tirol',
                      county: 'Innsbruck-Stadt',
                      city: 'Innsbruck',
                      postalCode: '6020',
                      additionalData: [
                        {
                          value: 'Österreich',
                          key: 'CountryName'
                        },
                        {
                          value: 'Tirol',
                          key: 'StateName'
                        },
                        {
                          value: 'Innsbruck-Stadt',
                          key: 'CountyName'
                        }
                      ]
                    }
                  }
                }
              ],
              viewId: 0
            }
          ]
        }
      },
      () => geocoder.geocode('Innsbruck')
    );

    const feature: GeocodingResult = result[0];
    expect(feature.name).toBe('Innsbruck, Tirol, Österreich');
    expect(feature.center).toStrictEqual({ lat: 47.268, lng: 11.3913 });
    expect(feature.bbox).toStrictEqual({
      _northEast: { lat: 47.35922, lng: 11.45587 },
      _southWest: { lat: 47.21082, lng: 11.30194 }
    });
    expect([[result]]).toMatchSnapshot();
  });
});

describe('L.Control.Geocoder.HEREv2', () => {
  it('geocodes Innsbruck', async () => {
    const geocodingParams = { at: '50.62925,3.057256' };
    const geocoder = new HEREv2({ apiKey: 'xxx', geocodingQueryParams: geocodingParams });
    const result = await mockFetchRequest(
      'https://geocode.search.hereapi.com/v1/discover?q=Innsbruck&apiKey=xxx&limit=10&at=50.62925%2C3.057256',
      {
        items: [
          {
            title: 'Salumeria Italiana',
            id: 'here:pds:place:840drt3p-898f6ee434794fe59895e71ccf9381e1',
            ontologyId: 'here:cm:ontology:restaurant',
            resultType: 'place',
            address: {
              label: 'Salumeria Italiana, 151 Richmond St, Boston, MA 02109, United States',
              countryCode: 'USA',
              countryName: 'United States',
              stateCode: 'MA',
              state: 'Massachusetts',
              county: 'Suffolk',
              city: 'Boston',
              district: 'North End',
              street: 'Richmond St',
              postalCode: '02109',
              houseNumber: '151'
            },
            position: { lat: 42.36355, lng: -71.05439 },
            access: [{ lat: 42.3635, lng: -71.05448 }],
            distance: 11,
            categories: [
              { id: '600-6300-0066', name: 'Grocery', primary: true },
              { id: '100-1000-0000', name: 'Restaurant' },
              { id: '100-1000-0006', name: 'Deli' },
              { id: '600-6300-0067', name: 'Specialty Food Store' }
            ],
            references: [
              { supplier: { id: 'core' }, id: '31213861' },
              { supplier: { id: 'tripadvisor' }, id: '3172680' },
              { supplier: { id: 'yelp' }, id: 'JNx0DlfndRurT-8KhSym7g' },
              { supplier: { id: 'yelp' }, id: 'P44VNcZUUNZfiFy-c4SUJw' }
            ],
            foodTypes: [
              { id: '304-000', name: 'Italian', primary: true },
              { id: '800-057', name: 'Pizza' },
              { id: '800-060', name: 'Sandwich' }
            ],
            contacts: [
              {
                phone: [
                  { value: '+16175234946' },
                  { value: '+16175238743' },
                  { value: '+16177204243' }
                ],
                fax: [{ value: '+16175234946' }],
                www: [{ value: 'http://www.salumeriaitaliana.com' }],
                email: [{ value: 'contact@salumeriaitaliana.com' }]
              }
            ],
            openingHours: [
              {
                text: ['Mon-Sat: 08:00 - 17:00', 'Sun: 10:00 - 16:00'],
                isOpen: false,
                structured: [
                  {
                    start: 'T080000',
                    duration: 'PT11H00M',
                    recurrence: 'FREQ:DAILY;BYDAY:MO,TU,WE,TH,FR,SA'
                  },
                  { start: 'T100000', duration: 'PT06H00M', recurrence: 'FREQ:DAILY;BYDAY:SU' }
                ]
              }
            ]
          }
        ]
      },
      () => geocoder.geocode('Innsbruck')
    );

    const feature: GeocodingResult = result[0];
    expect(feature.name).toBe(
      'Salumeria Italiana, 151 Richmond St, Boston, MA 02109, United States'
    );
    expect(feature.center).toStrictEqual({ lat: 42.36355, lng: -71.05439 });
    expect([[result]]).toMatchSnapshot();
  });
});
