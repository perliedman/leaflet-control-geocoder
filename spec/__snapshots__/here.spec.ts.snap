// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`L.Control.Geocoder.HERE geocodes Innsbruck 1`] = `
Array [
  Array [
    Array [
      Object {
        "bbox": Object {
          "_northEast": Object {
            "lat": 47.35922,
            "lng": 11.45587,
          },
          "_southWest": Object {
            "lat": 47.21082,
            "lng": 11.30194,
          },
        },
        "center": Object {
          "lat": 47.268,
          "lng": 11.3913,
        },
        "name": "Innsbruck, Tirol, Österreich",
        "properties": Object {
          "additionalData": Array [
            Object {
              "key": "CountryName",
              "value": "Österreich",
            },
            Object {
              "key": "StateName",
              "value": "Tirol",
            },
            Object {
              "key": "CountyName",
              "value": "Innsbruck-Stadt",
            },
          ],
          "city": "Innsbruck",
          "country": "AUT",
          "county": "Innsbruck-Stadt",
          "label": "Innsbruck, Tirol, Österreich",
          "postalCode": "6020",
          "state": "Tirol",
        },
      },
    ],
  ],
]
`;
